include/boost/hana/view.hpp

```diff
@@ -279,6 +344,22 @@ BOOST_HANA_NAMESPACE_BEGIN
     //////////////////////////////////////////////////////////////////////////
     template <>
     struct at_impl<hana::view_tag> {
+        // cartesian_product_element_view
+        template <typename Sequences, std::size_t i, typename N>
+        static constexpr decltype(auto)
+        apply(detail::cartesian_product_element_view_t<Sequences, i> view, N const&) {
+            using Indices = decltype(detail::make_cartesian_product_indices(view.sequences_));
+            return hana::at_c<Indices::indices_of(i)[N::value]>(
+                      hana::at_c<N::value>(view.sequences_));
+        }
+
+        // cartesian_product_view
+        template <typename Sequences, typename N>
+        static constexpr decltype(auto)
+        apply(detail::cartesian_product_view_t<Sequences> view, N const&) {
+          return detail::cartesian_product_element_view_t<Sequences, N::value>{view.sequences_};
```

Comment:
Before this point I wondered why you needed this `cartesian_product_element_view_t` type, but now I see.